File: src/authoring/flow_registry.py
"""The author-selectable frozen-flow surface (the thing Step 7 freezes).

A scenario file selects frozen ``Flow`` types by name; this module is the explicit
mapping from an authoring type name → the frozen class + its wiring/param shape.

**Explicit, not introspected — by design.** A stock-id field is a plain ``str``
alias at runtime (``StockId = NewType("StockId", str)``), so field-type
introspection cannot tell a wiring field from any other string field. More to the
point, this registry *is* the authoring contract — the declared set of frozen
primitives a scenario may compose, and the wiring names it exposes for each. It
mirrors frozen constructor signatures; the signatures it mirrors are frozen, so the
"duplication" is a stable, deliberately-curated public surface (frozen in Step 7),
not incidental drift.

Step 0 registered the standalone **Crew** flows (the composition anchor); the
post-roadmap **Tier-1 unfreeze** grew this to the nine standalone Power / Thermal /
ECLSS flows — the rest of the frozen flow set that *fits this shape*
(``docs/plans/post-roadmap-flow-registry-growth.md``). A flow type's **param set** (its
``param_set`` name, a key into :data:`PARAM_LOADERS`, or ``None`` for a param-free
flow) is a fixed fact of the class — it names *which frozen loader* produces the
flow's params object. "Which concrete param values" is the per-flow authoring choice
(Step 1): the named default committed file, or a **parameter pack** — a param file in
the same ``{value, unit, source}`` schema that the *same frozen loader* reads (so a
pack's values are validated by the frozen bounds/unit guards, never bypassing them).

**The biosphere is deliberately absent, for a STRUCTURAL reason** (not a calibration
one): ``Allocation`` takes a composite ``ctx: CarbonContext`` bundling four param
objects with four stock ids, plus ``pheno``/``alloc`` on top — which a flat
``wiring_fields`` tuple + a single ``param_set`` cannot express. It also needs the aux
accumulator, the shared ``co2_pool`` feedback var and the two-rate master-day driver,
all deferred. It wants a *frozen-compartment include*, not flow-type entries.

**Registered ≠ calibrated.** Selecting a frozen type here means the *rate law* is
frozen and literature-derived, **not** that its numbers are validated: every param
these flows read except ``crew``'s two is a ``TODO(cite)`` placeholder pending the
deferred validation gate. A scenario built only from frozen types carries no
``has_authored_kinetics`` marker — which means "no authored kinetics", **not**
"validated". See ``docs/authoring-reference.md``, "Frozen is not calibrated".
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from domains.crew.flows import (
    FoodMetabolism,
    OxygenConsumption,
    WaterBalance,
)
from domains.crew.loader import load_crew_params
from domains.eclss.flows import (
    CO2Scrubber,
    Condenser,
    CrewMetabolism,
    O2Makeup,
)
from domains.eclss.loader import load_eclss_params
from domains.power.flows import (
    LoadDraw,
    SelfDischarge,
    SolarCharge,
)
from domains.power.loader import load_charge_params, load_self_discharge_params
from domains.thermal.flows import HeatInput, RadiatorReject
from domains.thermal.loader import load_thermal_params
from simcore.flow import Flow


@dataclass(frozen=True)
class FlowTypeSpec:
    """How one authoring flow-type name lowers to a frozen ``Flow`` constructor.

    ``cls`` is the frozen flow dataclass (typed ``Callable[..., Flow]`` because
    ``Flow`` is a structural ``Protocol``, not a nominal base — each concrete
    dataclass is a callable that returns a ``Flow``-conforming instance);
    ``wiring_fields`` the exact set of constructor keyword fields that take a
    ``StockId`` (the interpreter requires the scenario's ``wiring`` dict to match
    this set exactly); ``param_set`` names the frozen loader that produces this
    flow's params object (a key into :data:`PARAM_LOADERS`), or ``None`` for a
    param-free flow. Every frozen flow shares the ``(id, priority, *wiring[,
    params])`` shape.
    """

    cls: Callable[..., Flow]
    wiring_fields: tuple[str, ...]
    param_set: str | None


# The author-selectable frozen-flow surface. Keys are stable authoring type names
# (ASCII, dotted-namespace, matching the flows' own ids by convention). Step 0: the
# three standalone Crew flows. Tier 1: the nine standalone Power / Thermal / ECLSS
# flows. Grouped by domain for reading; the interpreter and the manifest both key by
# name, so the source order here is inert.
#
# FORCED vs DONOR-CONTROLLED is the fact an author most needs from this table, because
# it decides how positivity holds (see docs/authoring-reference.md):
#   * FORCED flows read a **hardcoded module forcing name** through ``env.get`` — NOT
#     through wiring. The scenario must declare a forcing under that exact name, and
#     the flow is therefore NOT multi-instanceable under a `{bundle, prefix}` include
#     (a namespaced forcing key is unreachable — the documented crew boundary).
#     Positivity is by SIZING (well-fed), never structural.
#   * DONOR-CONTROLLED flows read a stock, so their draw self-limits — positivity is
#     structural, but only while `k·dt < 1`. The frozen `k`s carry implicit `dt`
#     assumptions; an author picks `dt`. See the per-flow notes below.
FLOW_TYPES: dict[str, FlowTypeSpec] = {
    "crew.oxygen_consumption": FlowTypeSpec(
        cls=OxygenConsumption,
        wiring_fields=("o2_store", "o2_consumed"),
        param_set=None,
    ),
    "crew.food_metabolism": FlowTypeSpec(
        cls=FoodMetabolism,
        wiring_fields=("food_store", "exhaled_co2", "fecal_waste"),
        param_set="crew",
    ),
    "crew.water_balance": FlowTypeSpec(
        cls=WaterBalance,
        wiring_fields=("water_store", "crew_humidity", "urine"),
        param_set="crew",
    ),
    # --- Power (ENERGY, J; rates in W = J/s) ---------------------------------
    # FORCED on ``solar_power`` (W). Three legs: the η_c split names the charge-
    # conversion loss as heat. At η_c = 1 the heat leg is exactly 0 (still emitted).
    "power.solar_charge": FlowTypeSpec(
        cls=SolarCharge,
        wiring_fields=("solar_source", "battery", "waste_heat"),
        param_set="charge",
    ),
    # FORCED on ``load_power`` (W). 100 % dissipative — positivity by well-fed sizing
    # (a constant load CAN over-draw an empty battery; the backstop would fire).
    "power.load_draw": FlowTypeSpec(
        cls=LoadDraw,
        wiring_fields=("battery", "waste_heat"),
        param_set=None,
    ),
    # DONOR-CONTROLLED (leak ∝ battery). k = 1.0e-8 /s ⇒ k·dt ≈ 3.6e-5 at dt = 3600 s:
    # safe across any plausible authored dt (k·dt < 1 needs dt < 1.0e8 s ≈ 3 years).
    "power.self_discharge": FlowTypeSpec(
        cls=SelfDischarge,
        wiring_fields=("battery", "waste_heat"),
        param_set="self_discharge",
    ),
    # --- Thermal (ENERGY, J) --------------------------------------------------
    # FORCED on ``heat_load`` (W). Heat → heat, no form change ⇒ two legs, no loss leg.
    "thermal.heat_input": FlowTypeSpec(
        cls=HeatInput,
        wiring_fields=("heat_source", "node"),
        param_set=None,
    ),
    # DONOR-CONTROLLED and NONLINEAR (Stefan-Boltzmann T⁴). Positivity is structural
    # only AT THE FLOOR (R → 0 as Q → 0); near equilibrium it is by SIZING — the frozen
    # heat_capacity C = 1.0e7 J/K is sized so τ = C/(4εσA·T_eq³) ≈ 65 steps at
    # dt = 3600 s. A much larger dt overshoots. The one registered flow whose per-step
    # graph contains a transcendental (`**4`) ⇒ Tier-2 cross-port (tiers.json).
    "thermal.radiator_reject": FlowTypeSpec(
        cls=RadiatorReject,
        wiring_fields=("node", "space"),
        param_set="thermal",
    ),
    # --- ECLSS (OXYGEN mol / CARBON mol / WATER kg) ---------------------------
    # FORCED on THREE names at once (``o2_consumption`` / ``co2_production`` /
    # ``h2o_production``). Six legs across three quantities, each balanced
    # independently. The cabin pools it wires are SINGLE-QUANTITY (cabin_co2 is CARBON
    # with the 1:1 default composition, NOT {carbon:1, oxygen:2}) — atomic coupling is
    # the Phase-6 seam, so an authored cabin must not annotate composition here.
    "eclss.crew_metabolism": FlowTypeSpec(
        cls=CrewMetabolism,
        wiring_fields=(
            "cabin_o2",
            "cabin_co2",
            "cabin_h2o",
            "metabolic_o2_sink",
            "metabolic_co2_source",
            "metabolic_h2o_source",
        ),
        param_set=None,
    ),
    # DONOR-CONTROLLED. k_scrub = 1.0e-3 /s is sized for dt = 60 s (k·dt = 0.06). THE
    # sharpest authored-dt trap: at dt = 3600 s, k·dt = 3.6 > 1 — the draw exceeds the
    # stock and the Euler backstop rations (silently, unless the author reads
    # `rationed`). See docs/authoring-reference.md, "The dt constraint".
    "eclss.co2_scrubber": FlowTypeSpec(
        cls=CO2Scrubber,
        wiring_fields=("cabin_co2", "co2_removed"),
        param_set="eclss",
    ),
    # DONOR-CONTROLLED. k_cond = 5.0e-4 /s, likewise sized for dt = 60 s (k·dt = 0.03);
    # k·dt = 1.8 > 1 at dt = 3600 s.
    "eclss.condenser": FlowTypeSpec(
        cls=Condenser,
        wiring_fields=("cabin_h2o", "humidity_condensate"),
        param_set="eclss",
    ),
    # DEMAND-CONTROLLED toward o2_setpoint (the one flow of this shape). Not clamped:
    # above the setpoint the frozen law goes NEGATIVE, reversing the flow's direction —
    # the frozen "above-setpoint venting clamp is a deferred seam" boundary, reachable
    # by an author who wires cabin_o2 above the frozen 10.0 mol setpoint.
    "eclss.o2_makeup": FlowTypeSpec(
        cls=O2Makeup,
        wiring_fields=("o2_supply", "cabin_o2"),
        param_set="eclss",
    ),
}


# Named frozen param loaders. Each takes an **optional path** (defaulting to the
# committed frozen param file). A flow references its set by name (``params: crew``
# → the loader's default file) or supplies a **pack** (``params: {pack: …}`` → the
# same loader called with the pack's path), so a pack's values flow through the
# frozen loader's schema/bounds/unit validation — a pack is a param file, not a way
# around the guards.
PARAM_LOADERS: dict[str, Callable[..., object]] = {
    "crew": load_crew_params,
    # The Power self-discharge rate set (P5.5). Registered for Step 2's
    # authored-kinetics anchor: a ``DeclarativeFlow`` re-expressing ``SelfDischarge``
    # reads its ``k`` from this frozen loader (``param("self_discharge_rate")``), so the
    # authored value is the frozen one — bit-identical — passing the frozen guards.
    "self_discharge": load_self_discharge_params,
    # --- Tier 1: the sets the newly-registered flow types name ----------------
    # Each is reachable BOTH as a frozen type's `param_set` and — since every loader
    # returns a flat dataclass of floats — as a `kinetics` rate's `param("…")` source,
    # exactly like `self_discharge`. So registering these also widens the authored-
    # kinetics surface: an authored rate may now read η_c, the radiator properties, or
    # the ECLSS gains and get the *frozen* value through the *frozen* guards.
    # charge: charge_efficiency (η_c).
    "charge": load_charge_params,
    # thermal: emissivity / radiator_area / heat_capacity / space_temperature.
    "thermal": load_thermal_params,
    # eclss: co2_scrub_rate / condense_rate / o2_makeup_gain / o2_setpoint.
    "eclss": load_eclss_params,
}


def load_param_set(param_set: str, pack_path: Path | None) -> object:
    """Load a flow type's params: the committed default, or a pack file.

    ``param_set`` is the flow type's :attr:`FlowTypeSpec.param_set`; ``pack_path``
    is ``None`` for the committed default or an already-resolved path to a pack file
    (a param file in the frozen loader's own ``{value, unit, source}`` schema). The
    pack is read by the *same frozen loader*, so its values are validated identically.
    """
    loader = PARAM_LOADERS[param_set]
    return loader() if pack_path is None else loader(pack_path)

File: rust/crates/authoring/src/flow_registry.rs
//! The author-selectable frozen-flow surface — the Rust mirror of Python
//! `authoring.flow_registry` (Phase 9, Step 4b).
//!
//! A scenario file selects frozen `Flow` types by name; this module is the explicit
//! mapping from an authoring type name → the frozen constructor + its wiring/param
//! shape, plus the named frozen param sets a `kinetics` rate reads. **Explicit, not
//! introspected** — the registry *is* the authoring contract Step 7 freezes.
//!
//! Step 4b registered the standalone **Crew** flows (the composition anchor), matching
//! the Python Step-0 surface; the post-roadmap **Tier-1** unfreeze grew both ports to
//! twelve by adding the nine standalone Power / Thermal / ECLSS flows
//! (`docs/plans/post-roadmap-flow-registry-growth.md`). Frozen param values arrive from
//! the Option-C [`domains::params`] constants (the hex-float bundle,
//! `sibling_params.txt`), **not** a re-parsed YAML — so the Rust interpreter reads no
//! param file. **Parameter packs (arbitrary `{value,unit,source}` files) are deferred in
//! the Rust port** (a Rust pack reader would re-run the frozen bounds/unit validation on
//! an arbitrary file, which no anchor exercises and Step 5's Godot payoff does not need —
//! see [`crate::interpreter`]); a `params: {pack: …}` reference is an interpret-time error.
//!
//! **This mirror must stay in lockstep with Python `authoring.flow_registry`.** A
//! Python-only registration is a broken contract, not a half-done one: the authoring
//! manifest freezes the *Python* surface, so nothing here fails until an anchor exercises
//! the missing arm (`test_crossport.py`). The three surfaces below that must move together
//! for every added type are the [`flow_type`] match, the [`build_frozen_flow`] match, and
//! the hand-maintained [`FLOW_TYPE_NAMES`] list.
//!
//! **Registered ≠ calibrated**: selecting a frozen type means the *rate law* is frozen,
//! not that its numbers are validated — every param reachable here except `crew`'s two is
//! a `TODO(cite)` placeholder. See `docs/authoring-reference.md`, "Frozen is not
//! calibrated".

use std::collections::BTreeMap;

use domains::crew::{CrewParams, FoodMetabolism, OxygenConsumption, WaterBalance};
use domains::eclss::{CO2Scrubber, Condenser, CrewMetabolism, EclssParams, O2Makeup};
use domains::params;
use domains::power::{ChargeParams, LoadDraw, SelfDischarge, SelfDischargeParams, SolarCharge};
use domains::thermal::{HeatInput, RadiatorReject, ThermalParams};
use simcore::flow::Flow;

use crate::errors::AuthoringError;

/// How one authoring flow-type name lowers to a frozen `Flow` constructor: the exact
/// set of wiring keyword fields (order-significant — it is the constructor's positional
/// order) and the frozen param set the flow's constructor needs (or `None`).
#[derive(Debug, Clone, Copy)]
pub struct FlowTypeSpec {
    /// The wiring field names, in constructor-argument order.
    pub wiring_fields: &'static [&'static str],
    /// The frozen param-set name this flow's constructor consumes, if any.
    pub param_set: Option<&'static str>,
}

/// Look up a frozen flow type by its authoring name (the `FLOW_TYPES` dict analogue).
pub fn flow_type(name: &str) -> Option<FlowTypeSpec> {
    match name {
        "crew.oxygen_consumption" => Some(FlowTypeSpec {
            wiring_fields: &["o2_store", "o2_consumed"],
            param_set: None,
        }),
        "crew.food_metabolism" => Some(FlowTypeSpec {
            wiring_fields: &["food_store", "exhaled_co2", "fecal_waste"],
            param_set: Some("crew"),
        }),
        "crew.water_balance" => Some(FlowTypeSpec {
            wiring_fields: &["water_store", "crew_humidity", "urine"],
            param_set: Some("crew"),
        }),
        // --- Tier 1: Power ---
        "power.solar_charge" => Some(FlowTypeSpec {
            wiring_fields: &["solar_source", "battery", "waste_heat"],
            param_set: Some("charge"),
        }),
        "power.load_draw" => Some(FlowTypeSpec {
            wiring_fields: &["battery", "waste_heat"],
            param_set: None,
        }),
        "power.self_discharge" => Some(FlowTypeSpec {
            wiring_fields: &["battery", "waste_heat"],
            param_set: Some("self_discharge"),
        }),
        // --- Tier 1: Thermal ---
        "thermal.heat_input" => Some(FlowTypeSpec {
            wiring_fields: &["heat_source", "node"],
            param_set: None,
        }),
        "thermal.radiator_reject" => Some(FlowTypeSpec {
            wiring_fields: &["node", "space"],
            param_set: Some("thermal"),
        }),
        // --- Tier 1: ECLSS ---
        "eclss.crew_metabolism" => Some(FlowTypeSpec {
            wiring_fields: &[
                "cabin_o2",
                "cabin_co2",
                "cabin_h2o",
                "metabolic_o2_sink",
                "metabolic_co2_source",
                "metabolic_h2o_source",
            ],
            param_set: None,
        }),
        "eclss.co2_scrubber" => Some(FlowTypeSpec {
            wiring_fields: &["cabin_co2", "co2_removed"],
            param_set: Some("eclss"),
        }),
        "eclss.condenser" => Some(FlowTypeSpec {
            wiring_fields: &["cabin_h2o", "humidity_condensate"],
            param_set: Some("eclss"),
        }),
        "eclss.o2_makeup" => Some(FlowTypeSpec {
            wiring_fields: &["o2_supply", "cabin_o2"],
            param_set: Some("eclss"),
        }),
        _ => None,
    }
}

/// The known authoring flow-type names (for the "unknown flow type" error message and
/// the Step-7 completeness surface).
///
/// Hand-maintained and **sorted** — it is not derived from [`flow_type`] (a Rust `match`
/// cannot be enumerated), so adding a type means editing three places in this file. The
/// `flow_type_names_all_resolve` test below closes the gap in one direction (every name
/// here resolves); the cross-port anchors close the other.
pub const FLOW_TYPE_NAMES: &[&str] = &[
    "crew.food_metabolism",
    "crew.oxygen_consumption",
    "crew.water_balance",
    "eclss.co2_scrubber",
    "eclss.condenser",
    "eclss.crew_metabolism",
    "eclss.o2_makeup",
    "power.load_draw",
    "power.self_discharge",
    "power.solar_charge",
    "thermal.heat_input",
    "thermal.radiator_reject",
];

/// The known frozen param-set names (`PARAM_LOADERS` keys — the sets a `kinetics`
/// rate's `param("…")` may read, and the sets a frozen flow type names).
pub const PARAM_SET_NAMES: &[&str] = &["charge", "crew", "eclss", "self_discharge", "thermal"];

/// Construct a frozen-`type` flow from its wiring (the interpreter has already checked
/// the wiring keys match `wiring_fields` exactly). Loads the flow's frozen params from
/// the Option-C constants where needed. An unknown type is an [`AuthoringError`].
pub fn build_frozen_flow(
    type_name: &str,
    id: &str,
    _priority: i64,
    wiring: &BTreeMap<String, String>,
) -> Result<Box<dyn Flow>, AuthoringError> {
    // Helper: fetch a wired stock id by field name (guaranteed present by the caller's
    // exact-match wiring check, but surface a clear error rather than panic).
    let wire = |field: &str| -> Result<String, AuthoringError> {
        wiring
            .get(field)
            .cloned()
            .ok_or_else(|| AuthoringError::new(format!("flow {id:?}: missing wiring {field:?}")))
    };
    match type_name {
        "crew.oxygen_consumption" => Ok(Box::new(OxygenConsumption::new(
            id.to_string(),
            wire("o2_store")?,
            wire("o2_consumed")?,
        ))),
        "crew.food_metabolism" => Ok(Box::new(FoodMetabolism::new(
            id.to_string(),
            wire("food_store")?,
            wire("exhaled_co2")?,
            wire("fecal_waste")?,
            crew_params(),
        ))),
        "crew.water_balance" => Ok(Box::new(WaterBalance::new(
            id.to_string(),
            wire("water_store")?,
            wire("crew_humidity")?,
            wire("urine")?,
            crew_params(),
        ))),
        // --- Tier 1: Power ---
        "power.solar_charge" => Ok(Box::new(SolarCharge::new(
            id.to_string(),
            wire("solar_source")?,
            wire("battery")?,
            wire("waste_heat")?,
            charge_params(),
        ))),
        "power.load_draw" => Ok(Box::new(LoadDraw::new(
            id.to_string(),
            wire("battery")?,
            wire("waste_heat")?,
        ))),
        "power.self_discharge" => Ok(Box::new(SelfDischarge::new(
            id.to_string(),
            wire("battery")?,
            wire("waste_heat")?,
            self_discharge_params(),
        ))),
        // --- Tier 1: Thermal ---
        "thermal.heat_input" => Ok(Box::new(HeatInput::new(
            id.to_string(),
            wire("heat_source")?,
            wire("node")?,
        ))),
        "thermal.radiator_reject" => Ok(Box::new(RadiatorReject::new(
            id.to_string(),
            wire("node")?,
            wire("space")?,
            thermal_params(),
        ))),
        // --- Tier 1: ECLSS ---
        "eclss.crew_metabolism" => Ok(Box::new(CrewMetabolism::new(
            id.to_string(),
            wire("cabin_o2")?,
            wire("cabin_co2")?,
            wire("cabin_h2o")?,
            wire("metabolic_o2_sink")?,
            wire("metabolic_co2_source")?,
            wire("metabolic_h2o_source")?,
        ))),
        "eclss.co2_scrubber" => Ok(Box::new(CO2Scrubber::new(
            id.to_string(),
            wire("cabin_co2")?,
            wire("co2_removed")?,
            eclss_params(),
        ))),
        "eclss.condenser" => Ok(Box::new(Condenser::new(
            id.to_string(),
            wire("cabin_h2o")?,
            wire("humidity_condensate")?,
            eclss_params(),
        ))),
        "eclss.o2_makeup" => Ok(Box::new(O2Makeup::new(
            id.to_string(),
            wire("o2_supply")?,
            wire("cabin_o2")?,
            eclss_params(),
        ))),
        _ => Err(AuthoringError::new(format!(
            "flow {id:?}: unknown flow type {type_name:?} (known: {FLOW_TYPE_NAMES:?})"
        ))),
    }
}

/// The frozen Crew params from the Option-C bundle (`domains::params::crew()`).
fn crew_params() -> CrewParams {
    params::crew()
}

/// The frozen Power charge param (η_c) from the Option-C bundle.
fn charge_params() -> ChargeParams {
    params::charge()
}

/// The frozen Power self-discharge param (k) from the Option-C bundle.
fn self_discharge_params() -> SelfDischargeParams {
    params::self_discharge()
}

/// The frozen Thermal radiator params from the Option-C bundle.
fn thermal_params() -> ThermalParams {
    params::thermal()
}

/// The frozen ECLSS control-loop params from the Option-C bundle.
fn eclss_params() -> EclssParams {
    params::eclss()
}

/// Flatten a named frozen param set to the `name -> f64` map an authored-`kinetics`
/// rate's `param("…")` reads (the Python `_kinetics_param_map` analogue). Values come
/// from the Option-C constants, so an authored rate constant *is* the frozen one —
/// bit-identical, passing the frozen bounds/unit guards. An unknown set (or a pack) is
/// an [`AuthoringError`].
pub fn kinetics_param_map(param_set: &str) -> Result<BTreeMap<String, f64>, AuthoringError> {
    let mut map = BTreeMap::new();
    match param_set {
        "crew" => {
            let p = params::crew();
            map.insert(
                "respired_carbon_fraction".to_string(),
                p.respired_carbon_fraction,
            );
            map.insert(
                "insensible_water_fraction".to_string(),
                p.insensible_water_fraction,
            );
        }
        "self_discharge" => {
            let p = params::self_discharge();
            map.insert("self_discharge_rate".to_string(), p.self_discharge_rate);
        }
        // --- Tier 1: the three sets the newly-registered flow types name. Each mirrors
        // its Python loader's dataclass field-for-field: the Python side flattens with
        // `asdict()`, so a missing key here would make an authored rate that resolves on
        // one port fail on the other.
        "charge" => {
            let p = params::charge();
            map.insert("charge_efficiency".to_string(), p.charge_efficiency);
        }
        "thermal" => {
            let p = params::thermal();
            map.insert("emissivity".to_string(), p.emissivity);
            map.insert("radiator_area".to_string(), p.radiator_area);
            map.insert("heat_capacity".to_string(), p.heat_capacity);
            map.insert("space_temperature".to_string(), p.space_temperature);
        }
        "eclss" => {
            let p = params::eclss();
            map.insert("co2_scrub_rate".to_string(), p.co2_scrub_rate);
            map.insert("condense_rate".to_string(), p.condense_rate);
            map.insert("o2_makeup_gain".to_string(), p.o2_makeup_gain);
            map.insert("o2_setpoint".to_string(), p.o2_setpoint);
        }
        _ => {
            return Err(AuthoringError::new(format!(
                "unknown param set {param_set:?} (known: {PARAM_SET_NAMES:?})"
            )));
        }
    }
    Ok(map)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flow_type_names_all_resolve() {
        // FLOW_TYPE_NAMES is hand-maintained beside two `match`es that cannot be
        // enumerated, so it can silently drift from them. This closes one direction: every
        // advertised name resolves in `flow_type` AND builds in `build_frozen_flow`. (The
        // other direction — a match arm missing from the list — is caught by the cross-port
        // anchors, which author files by name.)
        for name in FLOW_TYPE_NAMES {
            let spec = flow_type(name).unwrap_or_else(|| panic!("{name} has no flow_type"));
            let wiring: BTreeMap<String, String> = spec
                .wiring_fields
                .iter()
                .map(|f| ((*f).to_string(), format!("stock.{f}")))
                .collect();
            build_frozen_flow(name, "test.flow", 0, &wiring)
                .unwrap_or_else(|e| panic!("{name} failed to build: {e:?}"));
        }
    }

    #[test]
    fn flow_type_names_is_sorted_and_matches_param_set_names() {
        // Sorted so the list reads as a set and diffs cleanly; every param set a flow type
        // names must be a known set (the Python `param_set in PARAM_LOADERS` mirror).
        let mut sorted = FLOW_TYPE_NAMES.to_vec();
        sorted.sort_unstable();
        assert_eq!(sorted, FLOW_TYPE_NAMES.to_vec());
        let mut sets = PARAM_SET_NAMES.to_vec();
        sets.sort_unstable();
        assert_eq!(sets, PARAM_SET_NAMES.to_vec());
        for name in FLOW_TYPE_NAMES {
            if let Some(set) = flow_type(name).and_then(|s| s.param_set) {
                assert!(PARAM_SET_NAMES.contains(&set), "{name} names unknown set {set}");
            }
        }
    }

    #[test]
    fn every_param_set_flattens() {
        // Each named set must flatten to a non-empty map — the authored-`kinetics`
        // `param("…")` path. An unknown set is an error, not a silent empty map.
        for set in PARAM_SET_NAMES {
            assert!(!kinetics_param_map(set).expect("known set").is_empty(), "{set}");
        }
        assert!(kinetics_param_map("nope").is_err());
    }

    #[test]
    fn authored_radiator_reject_reads_the_frozen_thermal_params() {
        // The one sliver no other gate can see. `thermal.radiator_reject` is excluded from
        // the bit-exact cross-port RUN comparison (it is Tier-2 — `powf`), and the
        // cross-port GRAPH DUMP that covers it does not render params. So a registry arm
        // that wired the WRONG param set would reach no gate at all.
        //
        // Compares the authored flow against a directly-constructed one rather than
        // recomputing Stefan-Boltzmann here: duplicating the frozen physics in a test
        // would just be a second place to get it wrong, and the question is only "did the
        // registry hand it params::thermal()".
        use std::collections::HashMap;

        use domains::thermal::{build_thermal, ThermalScenario};
        use simcore::environment::SourceResolver;

        // node0 far from the floor: at the frozen EQUILIBRIUM_SCENARIO's node0 = 0 the
        // radiator emits exactly 0 (T = T_space), which every param set agrees on — the
        // test would pass vacuously.
        let scenario = ThermalScenario {
            node0: 1.0e9,
            heat_load_w: 3000.0,
            dt_seconds: 3600.0,
        };
        let p = params::thermal();
        let (state, _registry) = build_thermal(&p, &scenario).expect("frozen build");
        let resolver = SourceResolver::new(HashMap::new(), HashMap::new()).expect("resolver");
        let env = resolver.bind(&state, scenario.dt_seconds);

        let wiring: BTreeMap<String, String> = [
            ("node".to_string(), "thermal.node".to_string()),
            ("space".to_string(), "boundary.space".to_string()),
        ]
        .into_iter()
        .collect();
        let authored = build_frozen_flow("thermal.radiator_reject", "thermal.r", 0, &wiring)
            .expect("builds");
        let reference = RadiatorReject::new(
            "thermal.r".to_string(),
            "thermal.node".to_string(),
            "boundary.space".to_string(),
            params::thermal(),
        );

        let rejected = |f: &dyn Flow| -> f64 {
            f.evaluate(&state, &env, scenario.dt_seconds)
                .expect("evaluates")
                .legs
                .iter()
                .find(|l| l.stock == "boundary.space")
                .expect("space leg")
                .amount
        };
        let got = rejected(authored.as_ref());
        assert_eq!(got, rejected(&reference), "authored radiator must use params::thermal()");

        // Teeth: the comparison is actually sensitive to the params — a different
        // emissivity gives a different answer, so the equality above is not vacuous.
        let wrong = RadiatorReject::new(
            "thermal.r".to_string(),
            "thermal.node".to_string(),
            "boundary.space".to_string(),
            ThermalParams {
                emissivity: p.emissivity / 2.0,
                ..params::thermal()
            },
        );
        assert_ne!(got, rejected(&wrong));
        assert!(got != 0.0, "the node must be off the floor or this proves nothing");
    }
}
